src/status_im/native_module/impl/module.cljs

```diff
@@ -126,79 +127,57 @@
-
-;; TODO(rasom): temporal solution, should be fixed on status-go side
-(def check-raw-calls-interval 400)
-(def interval-between-calls 100)
-;; contains all calls to jail before with duplicates
-(def raw-jail-calls (atom '()))
-;; contains only calls that passed duplication check
-(def jail-calls (atom '()))
-
-(defn remove-duplicate-calls
-  "Removes duplicates by [jail path] keys, remains the last one."
-  [[all-keys calls] {:keys [jail-id path] :as call}]
-  (if (and (contains? all-keys [jail-id path])
-           (not (#{:subscription :preview} (last path))))
-    [all-keys calls]
-    [(conj all-keys [jail-id path])
-     (conj calls call)]))
-
-(defn check-raw-calls-loop!
-  "Only the last call with [jail path] key is added to jail-calls list
-  from raw-jail-calls"
-  []
-  (go-loop [_ nil]
-    (let [[_ new-calls] (reduce remove-duplicate-calls [#{} '()] @raw-jail-calls)]
-      (reset! raw-jail-calls '())
-      (swap! jail-calls (fn [old-calls]
-                          (concat new-calls old-calls))))
-    (recur (<! (timeout check-raw-calls-interval)))))
-
-(defn execute-calls-loop!
-  "Calls to jail are executed ne by one with interval-between-calls,
-  which reduces chances of response shuffling"
-  []
-  (go-loop [_ nil]
-    (let [next-call (first @jail-calls)]
-      (swap! jail-calls rest)
-      (when next-call
-        (execute-call next-call)))
-    (recur (<! (timeout interval-between-calls)))))
-
-(check-raw-calls-loop!)
-(execute-calls-loop!)
+     #(do
+        (log/debug :call-jail :jail-id jail-id)
+        (log/debug :call-jail :path path)
+        ;; this debug message can contain sensitive info
+        #_(log/debug :call-jail :params params)
+        (let [params' (update params :context assoc
+                              :debug js/goog.DEBUG
+                              :locale rn-dependencies/i18n.locale)
+              cb      (fn [r]
+                        (let [{:keys [result] :as r'} (types/json->clj r)
+                              {:keys [messages]} result]
+                          (log/debug r')
+                          (doseq [{:keys [type message]} messages]
+                            (log/debug (str "VM console(" type ") - " message)))
+                          (callback r')))]
+          (.callJail status jail-id (types/clj->json path) (types/clj->json params') cb))))))
+
+(def ^:private queue-flush-time 400)
+
+(def ^:private call-queue (async/chan))
+(def ^:private deduplicated-calls (async/chan))
+
+(async-util/chunked-pipe! call-queue deduplicated-calls queue-flush-time)
+
+(defn- compare-calls
+  "Used as comparator deciding which calls should be de-duplicated.
+  Whenever we fetch suggestions, we only want to issue the last call
+  done in the `queue-flush-time` window, for all other calls, we have
+  to consider `:params` key as well."
+  [{:keys [jail-id path params]}]
+  (if (= :suggestions (last path))
+    [jail-id path]
+    [jail-id path params]))
+
+(go-loop []
```

Comment:
Could we control the lifecycle of this process (make it startable / stoppable / interruptible) ?

Seems like we might want to do that sometimes.

Reply:
I can't think of situation where it would be beneficial and doing it would actually introduce some additional complexity into the codebase.
All the module functionality is protocol impl anyway, so it's easy to swap it for different impl for example for the purpose of unit-testing.

Reply:
No, calling jail doesn't have anything to do with network the user is operating on, it works even without being connected at all.

Reply:
> I can't think of situation where it would be beneficial and doing it would actually introduce some additional complexity into the codebase.
probably when namespace is reloaded we would like to restart go-loop, in order to stop using outdated code inside it (if it was changed)

Reply:
Right but are they no scenario where a jail call could be network dependent so that it would make sense to purge / stop everything before switching network?

Reply:
we have 400ms delay here, i doubt that it's possible to switch networks so fast. I doubt that it's possible just to navigate to profile that fast :)

Reply:
Calls could be made async I guess? 
Sounds dangerous to me and much safer to cleanup everything before changing network. It could be a cause for pretty hard to debug issues.

Reply:
What does it change? As far as i understand currently jail cells will be reset when Login method is called, so even if some call will be that slow that it can survive navigation to Network Switching screen, we will not receive callback.

Reply:
Not expert on that part so I defer to you! Just wanted to raise that point it case it is an issue.